lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3OutputStream.java

```diff
@@ -0,0 +1,342 @@
+    }
+
+    @SuppressWarnings("NumericCastThatLosesPrecision")
+    @Override
+    public void write(int b)
+            throws IOException
+    {
+        oneByte[0] = (byte) b;
+        write(oneByte);
+    }
+
+    @Override
+    public void write(byte[] bytes, int offset, int length)
+            throws IOException
+    {
+        ensureOpen();
+
+        while (length > 0) {
+            ensureCapacity(length);
+
+            int copied = min(buffer.length - bufferSize, length);
+            arraycopy(bytes, offset, buffer, bufferSize, copied);
+            bufferSize += copied;
+
+            flushBuffer(false);
+
+            offset += copied;
+            length -= copied;
+        }
+    }
+
+    @Override
+    public void flush()
+            throws IOException
+    {
+        ensureOpen();
+        flushBuffer(false);
+    }
+
+    @Override
+    public void close()
+            throws IOException
+    {
+        if (closed) {
+            return;
+        }
+        closed = true;
+
+        if (failed) {
+            try {
+                abortUpload();
+                return;
+            }
+            catch (SdkException e) {
+                throw new IOException(e);
+            }
+        }
+
+        try {
+            flushBuffer(true);
+            memoryContext.close();
+            waitForPreviousUploadFinish();
+        }
+        catch (IOException | RuntimeException e) {
+            abortUploadSuppressed(e);
+            throw e;
+        }
+
+        try {
+            uploadId.ifPresent(this::finishUpload);
+        }
+        catch (SdkException e) {
+            abortUploadSuppressed(e);
+            throw new IOException(e);
+        }
+    }
+
+    private void ensureOpen()
+            throws IOException
+    {
+        if (closed) {
+            throw new IOException("Output stream closed: " + location);
+        }
+    }
+
+    private void ensureCapacity(int extra)
+    {
+        int capacity = min(partSize, bufferSize + extra);
+        if (buffer.length < capacity) {
+            int target = max(buffer.length, initialBufferSize);
+            if (target < capacity) {
+                target += target / 2; // increase 50%
+                target = constrainToRange(target, capacity, partSize);
+            }
+            buffer = Arrays.copyOf(buffer, target);
+            memoryContext.setBytes(buffer.length);
+        }
+    }
+
+    private void flushBuffer(boolean finished)
+            throws IOException
+    {
+        // skip multipart upload if there would only be one part
+        if (finished && !multipartUploadStarted) {
+            var request = PutObjectRequest.builder()
+                    .requestPayer(requestPayer)
+                    .bucket(location.bucket())
+                    .key(location.key())
+                    .contentLength((long) bufferSize)
+                    .applyMutation(builder -> {
+                        switch (sseType) {
```

Comment:
idea:  is it possible for our S3SseType cass to implement the mutation interface?  Then this would just be `.applyMutation(sseType)`